### PR TITLE
Preventing email leak in FIDO login and counting login attempts

### PR DESCRIPTION
**Pull request by jckong99**
On #60:
- Only attempts to hide email leaks from FIDO login, not any registration routes.
- Currently unable to test because, when logging in with an unregistered email, Windows notices that there are no credentials associated with it and does not offer biometric options, leaving only a security key as a sign-in method.

On #71:
- Password success/failure counts have locally been tested to be accurate.
- As described in [Balaji's comment](https://github.com/team-pass/FIDO-login/issues/71#issuecomment-1063269129), there does not appear to be an easy way to count FIDO failures, since, for example, bad fingerprint readings are localized to the FIDO authenticator and our server receives nothing. The code still implements attempt counting as if it worked.

### CODE REVIEW

**Comment by jckong99** [comment]
So, I forgot to create a new branch and accidentally made changes for #50 here. I feel like instructions were already kind of covered by the earlier addition of the "Daily Remaining Tasks" list on the user profile page, so I just extended that by including list items for registering any missing authentication options and adding a general statement to explain their purpose.

File: web/fido_app/models.py [hunk 1]
```diff
@@ -138,3 +138,19 @@ class Interaction(db.Model):
...
+class LoginAttempts(db.Model):
+    '''
+    Model associating email and date with login attempt counts,
+    both successful and not.
+    '''
+
+    id = db.Column(db.Integer, primary_key=True)
+
+    # Identifiers
+    email = db.Column(db.String(80), nullable=False)
```
Comment (zbreit) [inline web/fido_app/models.py:151]: Do we want to distinguish between failed FIDO vs password logins?

Comment (zbreit) [reply web/fido_app/models.py:151]: I know we can't accurately count all FIDO failures, but our model might want to account for this, at least. One option is just to add a `password_successes`, `password_failures`, `fido_successes`, and `fido_failures` fields

**Review by zbreit — CHANGES_REQUESTED** [review]
I just made a few suggestions. I also created a DB migration for your new table, which I'll push in a sec. Please let me know if you have any questions!

File: web/fido_app/webauthn_routes.py [hunk 2]
```diff
@@ -175,21 +175,21 @@ def webauthn_login_start():
...
-        allow_credentials=[PublicKeyCredentialDescriptor(id=credential_id_bytes)],
-    )
+    if user and user.credential_id:
+        credential_id_bytes = base64url_to_bytes(user.credential_id)
+        authentication_options = generate_authentication_options(
+            rp_id=RP_ID,
+            allow_credentials=[PublicKeyCredentialDescriptor(id=credential_id_bytes)],
+        )
+    else:
+        authentication_options = generate_authentication_options(
```
[the remark this answers is not in the document]
Comment (jckong99) [reply web/fido_app/webauthn_routes.py:187]: Added in [2e344d6](https://github.com/team-pass/FIDO-login/pull/72/commits/2e344d619a22ae193455b72059abfa8544599a11)

File: web/fido_app/routes.py [hunk 3]
```diff
@@ -39,18 +39,44 @@ def login():
...
     # Check that a user has a matching email/password combo
     user = User.query.filter_by(email=email).first()
 
     if user and not user.has_password():
+        # Update failed login count
+        attempts.failures = attempts.failures + 1
+        db.session.add(attempts)
+        db.session.commit()
+        
         flash('You have not configured a password for your account', 'error')
```
[the remark this answers is not in the document]
Comment (jckong99) [reply web/fido_app/routes.py]: Changed in [37b53ba](https://github.com/team-pass/FIDO-login/pull/72/commits/37b53ba267c298c30a48ddead5c3b8fb5c5d0bf8)

**Review by zbreit — APPROVED** [review]
LGTM

**Pull request merged**